src/ray/raylet/node_manager.cc

```diff
@@ -326,6 +326,30 @@ void NodeManager::Heartbeat() {
     heartbeat_data->add_resource_load_capacity(resource_pair.second);
   }
 
+  std::unordered_set<ObjectID> active_object_ids = worker_pool_.GetActiveObjectIDs();
+  uint64_t max_size = RayConfig::instance().raylet_active_object_ids_size();
+  if (active_object_ids.size() <= max_size) {
+    for (const auto &object_id : active_object_ids) {
+      if (heartbeat_data->active_object_id_size() == max_size) {
+        break;
+      }
+      std::string object_id_bytes = object_id.Binary();
+      heartbeat_data->add_active_object_id(object_id_bytes.data(),
+                                           object_id_bytes.size());
+    }
+  } else {
+    // If there are more than the configured maximum number of object IDs to send per
+    // heartbeat, sample from them randomly.
+    std::vector<ObjectID> active_object_id_vector(active_object_ids.begin(),
+                                                  active_object_ids.end());
+    std::random_shuffle(active_object_id_vector.begin(), active_object_id_vector.end());
+    for (const auto &object_id : active_object_ids) {
```

Comment:
I think you wanted to iterate over the first max_size entries from active_object_id_vector here, right?

Also there is no need to shuffle everything (there could be a lot of elements in active_object_id_vector), just sampling is sufficient.

Reply:
Oops, not sure how that happened...

As far as I can tell, this is the simplest way to do sampling from a vector prior to C++ 17 - if you have other suggestions let me know. Otherwise if you're very concerned about the performance we can implement our own.